### PR TITLE
NEWRELIC-5274 refactored system-info.js to reduce complexity

### PR DESCRIPTION
**Pull request by jmartin4563**
## Proposed Release Notes
Refactored `lib/system-info.js` to reduce complexity by switching to async/await instead of callbacks for async control

## Links
Closes NEWRELIC-5274

## Details
* Refactored `lib/system-info.js` to reduce complexity, specifically converted to using async/await from callbacks and breaking out logic branches into functions, main function `fetchSystemInfo` still accepts a callback to minimize change footprint
* Updated to use es2020 for eslint, need it for `?` operator and node14 has 95% compliance, should be safe
* Added unit tests to address drop in coverage from refactor
* Added JSDocs

### CODE REVIEW

File: lib/system-info.js [hunk 1]
```diff
@@ -1,219 +1,296 @@
...
-              if (!processorStats[key] || !isInteger(processorStats[key])) {
-                processorStats[key] = null
-              }
-            }
-
-            callback(processorStats)
-          }
-        )
-      })
-    })
```
Comment (mrickard) [inline lib/system-info.js:141]: (Applause gif)

**Review by mrickard — DISMISSED** [review]

File: lib/system-info.js [hunk 2]
```diff
@@ -1,219 +1,296 @@
...
-        logger.debug('No sysctl info found for names: %j', names)
-        callback(null)
+  for (const name of names) {
+    // returnValue being set means we already found what we were looking for, early exit for performance
+    if (returnValue) {
+      break
+    }
+
+    try {
+      const { stderr, stdout } = await exec(`sysctl -n ${name}`)
```
Comment (bizob2828) [inline lib/system-info.js:281]: i know you didn't write this but this is very unsafe and prone to command injections.  You should use `child_process.execFile('sysctl', ['-n', name])`

**Review by bizob2828 — CHANGES_REQUESTED** [review]

File: lib/system-info.js (hunk 2 above)
Comment (bizob2828) [reply lib/system-info.js:281]: Here's a decent example https://github.com/nodesecurity/eslint-plugin-security/blob/main/docs/avoid-command-injection-node.md

File: lib/system-info.js (hunk 1 above)
Comment (bizob2828) [reply lib/system-info.js:141]: _this_ is why promises and then async/await was added to Node.js

File: lib/system-info.js [hunk 3]
```diff
@@ -1,219 +1,296 @@
 /*
- * Copyright 2020 New Relic Corporation. All rights reserved.
+ * Copyright 2022 New Relic Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (jmartin4563) [reply lib/system-info.js:2]: That's my bad, I created a new file when I was refactoring so I could side-by-side it, then renamed when I was happy with the refactor, reverting

**Review by bizob2828 — DISMISSED** [review]

File: test/unit/system-info.test.js [hunk 4]
```diff
@@ -6,26 +6,407 @@
 'use strict'
 const tap = require('tap')
 const os = require('os')
-const systemInfo = require('../../lib/system-info')
+const proxyquire = require('proxyquire').noPreserveCache()
```
Comment (jmartin4563) [inline test/unit/system-info.test.js:9]: Refactoring exposed some lacking test coverage, so I added coverage for the two exported for testing purposes functions `_getProcessStats` and `_getMemoryStats`

**Review by bizob2828 — APPROVED** [review]

**Pull request merged**